### ISSUE TITLE
ContextExecutorService.wrap doesn’t propagate baggages 

### ISSUE THREAD

**Issue by gayanper** [micrometer-metrics/micrometer#3908]
**Describe the bug**
Instrumentation with ContextExecutorService.wrap for ExecutorService doesn’t handle propagation of  baggage fields created using the Tracer. When observing logs i can only see those baggage field values in the thread the baggage field was created. 

**Environment**

 - Micrometer version 1.11.1
 - Micrometer registry prometheus
 - OS: macOS
 - Java version: 17
 - SpringBoot version 3.1.0
 
**To Reproduce**
How to reproduce the bug:

Will submit a sample code if needed. 

**Expected behavior**
The baggage fields must be propagated similar to traceid and spanId through threads and must be available in the log MDC. 

**Additional context**

**Comment by marcingrzejszczak** [comment]
Can you create a project that replicates this issue? I don't see which Tracer you are using and what is the Micrometer Tracing version.

**Comment by gayanper** [comment]
@marcingrzejszczak I'm using the versions that comes with the spring-boot parent. I will try to create a sample project which is similar to the proprietary project which this problem has been discovered and share it tomorrow.

**Comment by gayanper** [comment]
[micrometer-demo.zip](https://github.com/micrometer-metrics/micrometer/files/11754186/micrometer-demo.zip)

If you run this add and trigger the endpoint

```
curl --location --request POST 'http://localhost:8080/jobs'
```

You will see the following logs

```
2023-06-15T09:12:03.669+02:00 INFO [b21141cb8aa1211cedc881b55a091090, b2126da0e2ebf64c, 88e1ae3b-705d-42e2-ab22-4ed8a5427a87] 57848 --- [pool-3-thread-1] c.example.micrometerdemo.JobController   : Starting new Job
2023-06-15T09:12:03.681+02:00 INFO [b21141cb8aa1211cedc881b55a091090, b2126da0e2ebf64c, ] 57848 --- [pool-3-thread-3] c.example.micrometerdemo.JobController   : Step 2 finished.
2023-06-15T09:12:03.775+02:00 INFO [b21141cb8aa1211cedc881b55a091090, b2126da0e2ebf64c, ] 57848 --- [pool-3-thread-2] c.example.micrometerdemo.JobController   : Step 1 finished.
2023-06-15T09:12:03.796+02:00 INFO [b21141cb8aa1211cedc881b55a091090, b2126da0e2ebf64c, ] 57848 --- [pool-3-thread-3] c.example.micrometerdemo.JobController   : Step 3 finished.
2023-06-15T09:12:03.796+02:00 INFO [b21141cb8aa1211cedc881b55a091090, b2126da0e2ebf64c, 88e1ae3b-705d-42e2-ab22-4ed8a5427a87] 57848 --- [pool-3-thread-1] c.example.micrometerdemo.JobController   : Finished new Job
```

As you can see the nested threads which logs `Steps # finished` doesn't have the baggage `jobId`

**Comment by gayanper** [comment]
I think after debugging my own sample code a bit more, I figured out that the baggage needs to be created before we start the observation. Will test more and get back.

**Comment by marcingrzejszczak** [comment]
Yeah so we do write about this that you should register it yourself (https://github.com/micrometer-metrics/tracing/blob/main/micrometer-tracing/src/main/java/io/micrometer/tracing/contextpropagation/ObservationAwareSpanThreadLocalAccessor.java#L36-L52) and here in the docs (https://micrometer.io/docs/tracing#_context_propagation_with_micrometer_tracing) but we should do much better job at making this work - I've added an issue here in Boot (https://github.com/spring-projects/spring-boot/issues/35906). I will close this issue in favor of that one.
